### PR TITLE
🌱 Fix upgrade test (CoreDNS verification)

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:
WIP/test PR for now. similar to https://github.com/kubernetes-sigs/cluster-api/pull/4469 just without the kind upgrade

Requires https://github.com/kubernetes/test-infra/pull/21778 *before merge*

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: test/framework/deployment_helpers.go [hunk 1]
```diff
@@ -241,7 +241,11 @@ func WaitForDNSUpgrade(ctx context.Context, input WaitForDNSUpgradeInput, interv
 		if err := input.Getter.Get(ctx, client.ObjectKey{Name: "coredns", Namespace: metav1.NamespaceSystem}, d); err != nil {
 			return false, err
 		}
-		if d.Spec.Template.Spec.Containers[0].Image == "k8s.gcr.io/coredns:"+input.DNSVersion {
+		// coredns image name has changed. So for now we accept:
+		// k8s.gcr.io/coredns:1.7.0
+		// k8s.gcr.io/coredns/coredns:v1.7.0
+		// k8s.gcr.io/coredns/coredns:v1.8.0
+		if strings.HasSuffix(d.Spec.Template.Spec.Containers[0].Image, "coredns:"+input.DNSVersion) {
```
Comment (fabriziopandini) [inline test/framework/deployment_helpers.go:248]: nit: given that we are concerned concerned here is the version, let's check for ":" version suffix only.
Also what about changing the comment into
```suggestion
		// NOTE: coredns image name has changed over time (k8s.gcr.io/coredns, 
		// k8s.gcr.io/coredns/coredns), so we are checking only the version actually changed.
		if strings.HasSuffix(d.Spec.Template.Spec.Containers[0].Image, input.DNSVersion) {
```

Comment (JoelSpeed) [reply test/framework/deployment_helpers.go:248]: Looking at this you'd still want to have a colon prefix to the check to make sure you're absolutely getting the right version

Eg if we don't then version `11.7.0` could be allowed as well right?

Comment (sbueringer) [reply test/framework/deployment_helpers.go:248]: Sounds good to me.

**Comment by sbueringer** [comment]
Test was green: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/4470/pull-cluster-api-e2e-workload-upgrade-1-21-latest-main/1381904442139348992

Review findings addressed. I'll run it again with the fixes:
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main

I'll drop this commit https://github.com/kubernetes-sigs/cluster-api/pull/4470/commits/3fd677a1208d5ec2e6a65eaca0d1d9f10049e243 and squash after https://github.com/kubernetes/test-infra/pull/21778 has been merged

**Comment by sbueringer** [comment]
/test pull-cluster-api-e2e-workload-upgrade-1-21-latest-main

(only 7/10 nodes were ready before conformance test were started)

File: test/e2e/cluster_upgrade.go [hunk 2]
```diff
@@ -106,11 +106,12 @@ func ClusterUpgradeConformanceSpec(ctx context.Context, inputGetter func() Clust
...
-			ClusterProxy:                input.BootstrapClusterProxy,
-			Cluster:                     clusterResources.Cluster,
-			ControlPlane:                clusterResources.ControlPlane,
-			EtcdImageTag:                input.E2EConfig.GetVariable(EtcdVersionUpgradeTo),
-			DNSImageTag:                 input.E2EConfig.GetVariable(CoreDNSVersionUpgradeTo),
+			ClusterProxy: input.BootstrapClusterProxy,
+			Cluster:      clusterResources.Cluster,
+			ControlPlane: clusterResources.ControlPlane,
+			EtcdImageTag: input.E2EConfig.GetVariable(EtcdVersionUpgradeTo),
+			// FIXME: don't merge this just a hack to test until https://github.com/kubernetes/test-infra/pull/21778 is merged
```
Comment (fabriziopandini) [inline test/e2e/cluster_upgrade.go:113]: @sbueringer if you remove this we can get this PR merged

**Comment by sbueringer** [comment]
@fabriziopandini done

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**